server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2756,4 +2756,31 @@ private static void trimUnsafeCommits(EngineConfig engineConfig) throws IOExcept
         final long minRetainedTranslogGen = Translog.readMinTranslogGeneration(translogPath, translogUUID);
         store.trimUnsafeCommits(globalCheckpoint, minRetainedTranslogGen, engineConfig.getIndexSettings().getIndexVersionCreated());
     }
+
+    protected void verifyEngineBeforeIndexClosing() {
+        final long globalCheckpoint = translog.getLastSyncedGlobalCheckpoint();
+        final long maxSeqNo = localCheckpointTracker.getMaxSeqNo();
+        if (globalCheckpoint != maxSeqNo) {
+            throw new IllegalStateException("Global checkpoint [" + globalCheckpoint
+                + "] mismatches maximum sequence number [" + maxSeqNo + "] on index shard " + shardId);
+        }
+    }
+
+    @Override
+    public String prepareEngineBeforeIndexClosing(String syncId) throws IOException {
+        try (ReleasableLock ignored = writeLock.acquire()) {
+            ensureOpen();
+            syncTranslog(); // make sure that we persist the global checkpoint to translog checkpoint
+            verifyEngineBeforeIndexClosing();
+            // we can reuse the existing syncId if there was no indexing activity since the last synced-flush.
+            if (indexWriter.hasUncommittedChanges() == false && lastCommittedSegmentInfos.userData.containsKey(Engine.SYNC_COMMIT_ID)) {
+                syncId = lastCommittedSegmentInfos.userData.get(Engine.SYNC_COMMIT_ID);
+            }
+            final CommitId commitId = flush(true, true);
+            if (syncId != null) {
+                syncFlush(syncId, commitId);
```

Comment:
Maybe add a comment here on why this is safe? My head is still spinning on whether this is safe or not, but I currently think the cluster block and primary all permit is enough.

Also, I think we can assert that we hold all permits.